xgcm/test/test_grid_ufunc.py

```diff
@@ -81,6 +83,112 @@ def test_invalid_signatures(self, signature):
+            ...
+
+        assert ufunc.signature == "()->()"
+
+    def test_annotated_args(self):
+        # TODO test hints without annotations
+        # TODO test hints with annotations that don't conform to Xgcm
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center"]):
+            ...
+
+        assert ufunc.signature == "(X:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center,Y:center"]):
+            ...
+
+        assert ufunc.signature == "(X:center,Y:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[np.ndarray, "X:left"],
+            b: Gridded[np.ndarray, "Y:right"],
+        ):
+            ...
+
+        assert ufunc.signature == "(X:left),(Y:right)->()"
+
+    def test_annotated_return_args(self):
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:center"]:
+            ...
+
+        assert ufunc.signature == "()->(X:center)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:left,Y:right"]:
+            ...
+
+        assert ufunc.signature == "()->(X:left,Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Tuple[
+            Gridded[np.ndarray, "X:left"], Gridded[np.ndarray, "Y:right"]
+        ]:
+            ...
+
+        assert ufunc.signature == "()->(X:left),(Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[xr.DataArray, "X:center"]
+        ) -> Gridded[np.ndarray, "X:left"]:
+            ...
+
+        assert ufunc.signature == "(X:center)->(X:left)"
+
+    @pytest.mark.xfail
+    def test_invalid_arg_annotation(self):
+        with pytest.raises(
+            ValueError, match="Argument a has an invalid grid signature"
+        ):
+
+            @as_grid_ufunc()
+            def ufunc(a: Gridded[np.ndarray, "(X:Mars)"]):
+                ...
+
+    @pytest.mark.xfail
+    def test_invalid_return_arg_annotation(self):
+        with pytest.raises(
+            ValueError, match="Return argument(s) have an invalid grid signature"
+        ):
+
+            @as_grid_ufunc()
+            def ufunc() -> Gridded[np.ndarray, "(X:Venus)"]:
+                ...
+
+    def test_both_sig_kwarg_and_hints_given(self):
+        with pytest.raises(
+            ValueError, match="only one of either type hints or signature kwarg"
+        ):
+
+            @as_grid_ufunc(signature="(X:center)->(X:left)")
+            def ufunc(
+                a: Gridded[np.ndarray, "(X:center)"]
+            ) -> Gridded[np.ndarray, "(X:left)"]:
+                ...
+
+    def test_type_hint_as_numpy_ndarray(self):
+        # This should fail mypy
+        # TODO but it doesn't
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[str, "(X:center)"]):
+            assert isinstance(a, np.ndarray)
+
+        # This should pass mypy
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "(X:center)"]):
+            assert isinstance(a, np.ndarray)
```

Comment:
I don't understand why the first of these two doesn't fail `mypy`. This means right now you can put literally anything in the annotated type hint and mypy won't complain, which is not what is supposed to happen.